smart_open/s3.py

```diff
@@ -170,44 +172,58 @@ def __init__(self, s3_object, content_length, version_id=None):
         self._object = s3_object
         self._content_length = content_length
         self._version_id = version_id
+        self._body_reload = True
+        self._body = None
+        self._position = 0
         self.seek(0)
+        self._load_body()
 
     def seek(self, position):
         """Seek to the specified position (byte offset) in the S3 key.
-
         :param int position: The byte offset from the beginning of the key.
         """
+        if position == self._position:
+            return
+        # close the current streaming body, and prepare to open a new
+        self._body.close()
+        self._body_reload = True
         self._position = position
+
+    def _load_body(self):
+        """open a new streaming body start from the current position"""
         range_string = make_range_string(self._position)
         logger.debug('content_length: %r range_string: %r', self._content_length, range_string)
-
-        #
-        # Close old body explicitly.
-        # When first seek(), self._body is not exist. Catch the exception and do nothing.
-        #
-        try:
-            self._body.close()
-        except AttributeError:
-            pass
-
-        if position == self._content_length == 0 or position == self._content_length:
+        if self._position == self._content_length == 0 or self._position == self._content_length:
             #
             # When reading, we can't seek to the first byte of an empty file.
             # Similarly, we can't seek past the last byte.  Do nothing here.
             #
             self._body = io.BytesIO()
         else:
             self._body = _get(self._object, self._version_id, Range=range_string)['Body']
+        self._body_reload = False
 
     def read(self, size=-1):
         if self._position >= self._content_length:
             return b''
-        if size == -1:
-            binary = self._body.read()
+
+        if self._body_reload:
+            self._load_body()
+
+        try:
+            if size == -1:
+                binary = self._body.read()
+            else:
+                binary = self._body.read(size)
+        except IncompleteReadError:
+            # the underlying connection of streaming body was closed by peer
```

Comment:
Minor English fixup.

```suggestion
            # The underlying connection of the streaming body was closed by the remote peer.
```